lib/Scheduling/ASAPScheduler.cpp

```diff
@@ -0,0 +1,63 @@
+//===- ASAPScheduler.cpp - As-soon-as-possible list scheduler -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of an as-soon-as-possible list scheduler for acyclic problems.
```

Comment:
I think this loops forever in the presence of a cycle? I know this is sort of a toy to demonstrate the functionality, but should it detect cycles and fail if it finds one? I think if you kept a set of seen `Operator *`s, or something like that, this wouldn't be too crazy to add.

Reply:
Fail-safe added.